src/eth/DSProxyService.js

```diff
@@ -0,0 +1,110 @@
+import PrivateService from '../core/PrivateService';
+import { Contract } from 'ethers';
+import { dappHub } from '../../contracts/abis';
+import { contractInfo } from '../../contracts/networks';
+
+export default class DSProxyService extends PrivateService {
+  constructor(name = 'proxy') {
+    super(name, ['web3']);
+  }
+
+  async authenticate() {
+    this._currentProxy = await this.getProxyAddress();
+  }
+
+  _proxyRegistry() {
+    return new Contract(
+      this._registryInfo().address,
+      this._registryInfo().abi,
+      this.get('web3')
+        .ethersProvider()
+        .getSigner()
+    );
+  }
+
+  _registryInfo() {
+    return contractInfo(this._network()).PROXY_REGISTRY[0];
+  }
+
+  _network() {
+    switch (this.get('web3').networkId()) {
+      case 1:
+        return 'mainnet';
+      case 42:
+        return 'kovan';
+      case 999:
+        return 'test';
+    }
+  }
+
+  _setCurrentProxy(transaction) {
+    return new Promise(async resolve => {
+      await transaction;
+      resolve(await this.getProxyAddress());
+    });
+  }
+
+  _resetDefaults(newProxy) {
+    this._currentProxy = newProxy;
+    this._currentAccount = this.get('web3').currentAccount();
+  }
+
+  currentProxy() {
+    return this._currentAccount === this.get('web3').currentAccount()
+      ? this._currentProxy
+      : this.getProxyAddress();
+  }
+
+  async build() {
+    const transaction = this._proxyRegistry().build();
+    this._currentProxy = await this._setCurrentProxy(transaction);
```

Comment:
then you can toss `_setCurrentProxy entirely

Reply:
I did this so I could return the transaction, but revisiting it here made me realize that I wasn't actually wrapping this in a transaction object like I intended to. So the updated version does that (and takes into account the nonce and transaction settings). I removed `_setCurrentProxy`, but as of now, I'm still returning the transaction object.

This may be relevant for instance in the case of Oasis Direct, where many of the build and execute functions are actually just one call to `build()` followed by a second call to `execute()`. Thoughts on this @levity ?